# Bike and Income Data SQL Queries

This notebook connects to the PostgreSQL database and runs a set of analytical queries on:
- `income_clean`
- `bicycle_station_status_clean`

It saves outputs as CSVs and previews each result.


In [ ]:
import pandas as pd
from sqlalchemy import create_engine
from pathlib import Path
import geopandas as gpd


# DB credentials
DB_PARAMS = {
    "host": "dtim.essi.upc.edu",
    "port": 5432,
    "dbname": "dbakosschneider",
    "user": "akosschneider",
    "password": "DMT2025!"
}

def get_engine():
    return create_engine(
        f"postgresql+psycopg2://{DB_PARAMS['user']}:{DB_PARAMS['password']}"
        f"@{DB_PARAMS['host']}:{DB_PARAMS['port']}/{DB_PARAMS['dbname']}"
    )

engine = get_engine()
print("✅ Connected to database")


✅ Connected to database


In [2]:
def run_query(query: str):
    try:
        df = pd.read_sql_query(query, engine)
        print(f"✅ Success: {len(df)} rows")
        display(df.head())
        return df
    except Exception as e:
        print(f"❌ Query failed: {e}")
        return pd.DataFrame()


### Query: `avg_income_per_district`

In [3]:
df_avg_income_per_district = run_query("""SELECT
            codi_districte,
            AVG(import_euros) AS avg_income
        FROM income_clean
        GROUP BY codi_districte
        ORDER BY avg_income DESC;""")

✅ Success: 10 rows


codi_districte    avg_income
0               5  33351.327682
1               4  27844.956115
2               2  23998.948175
3               6  23100.095846
4              10  19190.856569

### Query: `avg_income_over_years`

In [4]:
df_avg_income_over_years = run_query("""SELECT
            EXTRACT(YEAR FROM "any") AS year,
            ROUND(AVG(import_euros)::NUMERIC, 2) AS avg_income
        FROM income_clean
        GROUP BY year
        ORDER BY year;""")

✅ Success: 7 rows


year  avg_income
0  2015.0    20214.79
1  2016.0    20875.86
2  2017.0    21091.32
3  2018.0    21390.68
4  2019.0    22467.19

### Query: `income_distribution_values`

In [5]:
df_income_distribution_values = run_query("""SELECT import_euros
        FROM income_clean
        WHERE import_euros IS NOT NULL;""")

✅ Success: 7476 rows


import_euros
0      14250.55
1      16005.85
2      16190.46
3      17573.10
4      16612.03

### Query: `income_by_tract_centroid`

In [6]:
df_income_by_tract_centroid = run_query("""SELECT
            seccio_censal,
            ST_Centroid(geometry) AS centroid,
            AVG(import_euros) AS avg_income
        FROM income_clean
        WHERE geometry IS NOT NULL
        GROUP BY seccio_censal, geometry;""")

✅ Success: 1068 rows


seccio_censal                                           centroid  \
0           1001  0101000020E6100000AAF5BCC91B5D01405F2DFE0E84B0...   
1           1002  0101000020E6100000AAF5BCC91B5D01405F2DFE0E84B0...   
2           1003  0101000020E6100000AAF5BCC91B5D01405F2DFE0E84B0...   
3           1004  0101000020E6100000AAF5BCC91B5D01405F2DFE0E84B0...   
4           1005  0101000020E6100000AAF5BCC91B5D01405F2DFE0E84B0...   

     avg_income  
0  13573.477143  
1  10611.092857  
2  10734.057143  
3  13472.698571  
4  11111.730000

### Query: `bike_station_availability_stats`

In [7]:
df_bike_station_availability_stats = run_query("""SELECT
            station_id,
            AVG(num_bikes_available) AS avg_bikes_available,
            MIN(num_bikes_available) AS min_bikes,
            MAX(num_bikes_available) AS max_bikes
        FROM bicycle_station_status_clean
        GROUP BY station_id
        ORDER BY avg_bikes_available DESC;""")

✅ Success: 521 rows


station_id  avg_bikes_available  min_bikes  max_bikes
0         529           198.000000        198        198
1         530           198.000000        198        198
2         180            29.194426          0         51
3          39            26.225997          0         44
4          46            24.831051          0         54

Bike Station Turnover (approximation for activity):

Estimation of how active each station is by looking at how many bikes are typically available.

In [8]:
bike_station_turnover = run_query("""SELECT
    station_id,
    AVG(num_bikes_available) AS avg_bikes_available,
    MAX(num_bikes_available) AS max_bikes,
    MIN(num_bikes_available) AS min_bikes,
    COUNT(*) AS observations
FROM bicycle_station_status_clean
GROUP BY station_id
ORDER BY avg_bikes_available DESC;""")



✅ Success: 521 rows


station_id  avg_bikes_available  max_bikes  min_bikes  observations
0         529           198.000000        198        198           102
1         530           198.000000        198        198             9
2         180            29.194426         51          0        139801
3          39            26.225997         44          0        139971
4          46            24.831051         54          0        139971

Stations 529 and 530: Very high constant availability (198 bikes) and with no variation (min = max = avg) and they have very low observation count. Maybe these stations are test, virtual ones or inactive stations.

Stations 180, 39 and 46: Large historical data (139k+ status records). Expected variation how expected variation: bikes come and go. Useful for identifying busy areas (high variance → high turnover)

Where are these stations?
SUGGESTION: we can relate these ones with the income and population. Maybe height? And maybe we could delete these two station that doesn't seem to be real.

In [9]:

bike_station_ubi = run_query("""SELECT
    station_id,
    name,
    ST_SetSRID(ST_MakePoint(lon, lat), 4326) AS geometry
FROM bicycle_station_information_clean
WHERE station_id IN (180, 39, 46);""")




✅ Success: 68650 rows


station_id                                    name  \
0          39                             PL.PAU VILA   
1          46               C/ RAMON TRIAS FARGAS, 19   
2         180  GRAN VIA DE LES CORTS CATALANES/MOSSÈN   
3          39                             PL.PAU VILA   
4          46               C/ RAMON TRIAS FARGAS, 19   

                                            geometry  
0  0101000020E61000007BBFD18E1B7E0140C422861DC6B0...  
1  0101000020E6100000106F42C6488B01400B0852CEBCB1...  
2  0101000020E6100000C0D3BF7F4E1C01408F33F21E0CAF...  
3  0101000020E61000007BBFD18E1B7E0140C422861DC6B0...  
4  0101000020E6100000106F42C6488B01400B0852CEBCB1...


Station usage over time:

Visualize bike availability trends per station per month

In [10]:
bike_station_usage = run_query("""SELECT
    station_id,
    DATE_TRUNC('month', last_updated)::date AS month,
    AVG(num_bikes_available) AS avg_bikes
FROM bicycle_station_status_clean
GROUP BY station_id, DATE_TRUNC('month', last_updated)
ORDER BY station_id, DATE_TRUNC('month', last_updated);""")

✅ Success: 15907 rows


station_id       month  avg_bikes
0           1  2019-03-01  24.278618
1           1  2019-04-01  20.652377
2           1  2019-05-01  20.480432
3           1  2019-06-01  19.837465
4           1  2019-07-01  15.358115

Create a new layer 

In [3]:
from sqlalchemy import create_engine

engine = create_engine(
    "postgresql+psycopg2://akosschneider:DMT2025!@dtim.essi.upc.edu:5432/dbakosschneider"
)


In [ ]:
import geopandas as gpd

query = '''
SELECT
    s1.station_id AS origin_station,
    s2.station_id AS destination_station,
    ST_SetSRID(ST_MakeLine(
        ST_MakePoint(s1.lon, s1.lat),
        ST_MakePoint(s2.lon, s2.lat)
    ), 4326) AS geometry
FROM bicycle_station_information_clean s1
JOIN bicycle_station_information_clean s2
  ON s1.station_id != s2.station_id
WHERE s1.station_id IN (180, 39, 46)
  AND s2.station_id IN (180, 39, 46);
'''
od_lines = gpd.read_postgis(query, engine, geom_col="geometry")
od_lines